tools/testrunner/main.go

```diff
@@ -0,0 +1,340 @@
+/*
+Start the test runner (gotestsum) as usual.
+
+Download https://github.com/databricks/cli/blob/ciconfig/known_failures.txt in parallel.
+
+If download was successful by the time test runner finishes and test runner finishes with non-zero code,
+analyze failures in the test output identified by --jsonfile option. If all failures are expected (listed in known_failures.txt)
+then the failure is masked, the process exits with 0.
+*/
+package main
+
+import (
+	"bufio"
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"os"
+	"os/exec"
+	"strings"
+	"sync/atomic"
+	"syscall"
+)
+
+const (
+	maxConfigSize = 10 * 1024 // 10KB
+	repoConfigURL = "https://raw.githubusercontent.com/databricks/cli/ciconfig/known_failures.txt"
+	cutPrefix     = "github.com/databricks/cli/"
+)
+
+type TestResult struct {
+	Action  string `json:"Action,omitempty"`
+	Package string `json:"Package,omitempty"`
+	Test    string `json:"Test,omitempty"`
+}
+
+func getExitCode(err error) (int, error) {
+	if err == nil {
+		return 0, nil
+	}
+	if exitError, ok := err.(*exec.ExitError); ok {
+		if status, ok := exitError.Sys().(syscall.WaitStatus); ok {
+			return status.ExitStatus(), nil
+		}
+	}
+	return 1, err
+}
+
+func main() {
+	if len(os.Args) < 2 {
+		fmt.Fprintf(os.Stderr, "Usage: %s <command> [args...]\n", os.Args[0])
+		os.Exit(1)
+	}
+
+	// Find --jsonfile argument
+	jsonFile := ""
+	for i, arg := range os.Args {
+		if arg == "--jsonfile" && i+1 < len(os.Args) {
+			jsonFile = os.Args[i+1]
+			break
+		}
+	}
+
+	if jsonFile == "" {
+		fmt.Println("No --jsonfile argument found")
+	}
+
+	// Start the main subprocess
+	cmd := exec.Command(os.Args[1], os.Args[2:]...)
+	cmd.Stdout = os.Stdout
+	cmd.Stderr = os.Stderr
+	cmd.Stdin = os.Stdin
+
+	var configContent atomic.Value
+
+	// Start background config download
+	go func() {
+		content, err := downloadConfig(context.Background())
+		if err != nil {
+			fmt.Printf("testrunner: Failed to download %s: %v\n", repoConfigURL, err)
+		} else {
+			configContent.Store(content)
+		}
+	}()
+
+	// Run the main command
+	err := cmd.Run()
+
+	exitCode, err := getExitCode(err)
+	if err != nil {
+		fmt.Printf("testrunner: Failed to run command: %v\n", err)
+		os.Exit(1)
+	}
+
+	// Success case, exit early
+	if exitCode == 0 || jsonFile == "" {
+		os.Exit(exitCode)
+	}
+
+	// Check if config is ready
+	content := configContent.Load()
```

Comment:
The content download is expected to be much faster than running the tests. Should we block on downlonading the content always to keep this simpler?

Reply:
Sure, usually test run with finish first. However, It'd be annoying to be blocked by this even if rare.